### ISSUE TITLE
Different results for stream updates across clay and mainnet

### ISSUE THREAD

**Issue by mikehroth** [ceramicnetwork/js-ceramic#1796]
# Description

Tile update behavior appears to differ from Clay to the Mainnet/ELP when performing consecutive updates on the same stream. On testnet we see 3 commits after a create and two updates to the same stream. On mainnet there is only one.

# Technical Information

```javascript
const { TileDocument } = require('@ceramicnetwork/stream-tile');
const storage = require('@fungyproof/storage')
const HDWalletProvider = require('@truffle/hdwallet-provider');

const store = function() {
    const db = {};

    return {
        get: function(key) {
            return Promise.resolve(db[key]);
        },
        set: function(key, val) {
            db[key] = val;
        }
    }
}

// Testnet / Clay
function testClay() {
    process.env.CERAMIC_NODE =  'https://ceramic-clay.3boxlabs.com'
    const cm = new storage.CeramicProvider(new HDWalletProvider(
       "YOUR SEED PHRASE",
        "YOUR RPC NODE" // ropsten
    ), store())

    cm.connect().then(function(){
        
        // empty tile
        cm.createTile({}).then(function(t){

            const id = t.id.toString()

            // first update
            cm.updateTile(id, { updated: 'updated' }).then(function(t){

                // second update
                cm.updateTile(id, { updated: 'changed', final: { final: 'state' } }).then(function(t){
                    console.log('done', id)

                    // print commits
                    TileDocument.load(cm._ceramic, id).then(function(tile) {
                        console.log('TESTNET: commits', tile.allCommitIds)
                    });

                })

            })

        })

    })
}


// Mainnet
function testMainnet() {
    process.env.CERAMIC_NODE =  'https://fungy.link' // mainnet node
    const cm = new storage.CeramicProvider(new HDWalletProvider(
       "YOUR SEED PHRASE",
        "YOUR RPC NODE" // mainnet
    ), store())

    cm.connect().then(function(){
        
        // empty tile
        cm.createTile({}).then(function(t){

            const id = t.id.toString()

            // first update
            cm.updateTile(id, { updated: 'updated' }).then(function(t){

                // second update
                cm.updateTile(id, { updated: 'changed', final: { final: 'state' } }).then(function(t){
                    console.log('done', id)

                    // print commits
                    TileDocument.load(cm._ceramic, id).then(function(tile) {
                        console.log('MAINNET: commits', tile.allCommitIds)
                    });

                })

            })

        })

    })
}

testClay()
testMainnet()
```

**Results**

```
done kjzl6cwe1jw14blitmu3ezy1r57pyfnzozcltrihhadno5v9odxvgxssjif1ub8
MAINNET: totalCommits [
  CommitID(k3y52l7qbv1fryqh1uqe0anm4h12wkveltndn7nngaxt4btqctf4vsgcqyqz13jsw)
]
done kjzl6cwe1jw149c5gbw2swr4fgso0fnnfbfcbtjfysqv3wiqxv7ezidilc10x1w
TESTNET: totalCommits [
  CommitID(k3y52l7qbv1fryaees4k3xkwvhzfur3c6kx93o229jgf17r9bktwqkio47pjaj1fk),
  CommitID(k1dpgaqe3i64kjv7g367txqeucmn339etpyjhdytzdvou5vasugvh6pkaxugtpz5ayhemdz1mraquhcs93h13e2e8gno19qbx4hj6l5ppqvo3nul38z970818),
  CommitID(k1dpgaqe3i64kjv7g367txqeucmn339etpyjhdytzdvou5vasugvh6pkaxugtpz5ayhemdywybf90cpdkawpqc8k8l8jnladicaghzivy3sgypaly2xjbubyz)
]
```

**Comment by mikehroth** [comment]
I've made some progress. The function I was using in our custom ceramic provider was first retrieving the tile via `TileDocument.load`, then performing the update. If I retrieve the tile and update directly the commits do appear to show up. The issue now is that while they appear in the immediate log, they do not appear to actually propagate to the network. 

Which means if I attempt to load the stream later on, the only commit that will show up is the original `create` commit.
Further, if I retrieve this stream and perform an update later, nothing happens at all - no new commits.

```javascript
// Mainnet
function testMainnet() {
    process.env.CERAMIC_NODE =  'https://fungy.link' // mainnet node
    const cm = new storage.CeramicProvider(new HDWalletProvider(
        "SEED PHRASE",
        `Mainnet RPC`
    ), store())

    cm.connect().then(function(){

        //  { deterministic: true }
        TileDocument.create(cm._ceramic, {}).then(function(tile){
            console.log('created', tile.id.toString())
            console.log(tile.content, tile.allCommitIds);

            tile.update({ updated: 'updated' }, {}, {
                anchor: true,
                publish: true
            }).then(function() {
                console.log(tile.content, tile.allCommitIds);

                tile.update({ updated: 'final', more: { content: 'content' } }, {}, {
                    anchor: true,
                    publish: true
                }).then(function(){
                    console.log(tile.id.toString())
                    console.log(tile.content, tile.allCommitIds)
                })
            })

        })

    })
}
```

**Comment by stbrody** [comment]
The first code example you posted relies on lots of custom support functions whose source isn't included here, so it's hard to see if there's anything wrong with the code.  The second code sample that you posted in your more recent comment looks right, and seems to be working as you say, so it's hard to say.

> The issue now is that while they appear in the immediate log, they do not appear to actually propagate to the network.

How are you testing that they are propagating to the network?

>  if I attempt to load the stream later on, the only commit that will show up is the original create commit.

I notice in your code example that you aren't pinning the streams you are creating.  If you [pin the stream while creating it](https://developers.ceramic.network/build/javascript/pinning/#pin-a-stream-while-creating-it), does that change the behavior at all?

**Comment by stbrody** [comment]
also note that the ability to pin a stream while creating it only exists in newer versions of Ceramic, so make sure to update your node to the newest version first.

**Comment by mikehroth** [comment]
> The first code example you posted relies on lots of custom support functions whose source isn't included here, so it's hard to see if there's anything wrong with the code. The second code sample that you posted in your more recent comment looks right, and seems to be working as you say, so it's hard to say.
> 
> > The issue now is that while they appear in the immediate log, they do not appear to actually propagate to the network.
>>>You're right, and in-fact there was a bug specifically related to this code which i'll explain below.
> 
> How are you testing that they are propagating to the network?
> 
> > if I attempt to load the stream later on, the only commit that will show up is the original create commit.
>>> I was pulling the streams from a different gateway than our own to make sure they would show up.

> I notice in your code example that you aren't pinning the streams you are creating. If you [pin the stream while creating it](https://developers.ceramic.network/build/javascript/pinning/#pin-a-stream-while-creating-it), does that change the behavior at all?
>> I stripped that out for brevity but we are pinning all of our streams and we are doing it in one call on the latest version of js-ceramic.

**Comment by mikehroth** [comment]
@stbrody and @oed I've resolved all of our node issues at this point, I'll post our issues/resolutions here for the sake of posterity in case anyone else runs across this.

Our issues were 3 fold:

1) The code that we were using to create the streams contained a small bug which could occasionally set the `controller` as null, this then prevented future updates to the same stream from succeeding / propagating to the network.

2) Our nodes were running out of CPU units and occasionally returning 502s, failing to update the streams. This seemed to happen in general when there was more than 3 commits. We increased our CPU/Mem to 2048units / 4096 mem

3) Using the S3 storage for our storage state can very occasional throw a 502 - there's no great way to handle this afaik for now. If it becomes a problem as we scale we may need to store Ceramic and IPFS state on disk then backup state to external solutions such as Filecoin/S3.

**Comment by stbrody** [comment]
Hi @mikehroth, thank you for the update.  I have filed https://github.com/ceramicnetwork/js-ceramic/issues/1816 which should prevent #1 from happening again, and I put a project on our backlog to increase the resiliency of the S3 state store in the face of transient errors which should help with #3.  For #2 it'll be a longer-term series of work to reduce the resource consumption that Ceramic needs, but there is lots of room for optimizations here and this should improve over time.

Sorry for the inconvenience and trouble you faced!  And thanks again for the detailed and helpful report of the issues you encountered so we can improve things for the future!

If there are no objections, I will go ahead and close this ticket now as it seems the original issue has been addressed and there are tickets or projects to capture the remaining improvements.
